### PR TITLE
Replace underlying Region in Scala

### PR DESCRIPTION
**Pull request by catoverdrive**
rewrite Region.scala to build on the RegionPool backed region.

### CODE REVIEW

File: hail/src/main/scala/is/hail/annotations/Region.scala [hunk 1]
```diff
@@ -267,3 +260,25 @@ final class Region() extends NativeBase() {
   }
 
 }
+
+object RegionPool {
+  private val pools = new mutable.HashMap[Long, (Object, RegionPool)]()
+
+  private val contextLock = new Object()
+
+  private val theRegionPool: RegionPool = new RegionPool()
```
Comment (cseed) [inline hail/src/main/scala/is/hail/annotations/Region.scala]: unused

File: hail/src/main/scala/is/hail/annotations/Region.scala [hunk 2]
```diff
@@ -267,3 +260,25 @@ final class Region() extends NativeBase() {
...
+
+  def get: RegionPool = {
+    val (contextLock, rp) = pools.getOrElseUpdate(Thread.currentThread().getId(), (new Object, new RegionPool()))
```
Comment (cseed) [inline hail/src/main/scala/is/hail/annotations/Region.scala]: lock this (or no lock if thread safe)

**Comment by catoverdrive** [comment]
currently, all the c++ code is acquiring things called "ScalaRegions" to minimize the amount of changes in this PR. One of the next steps will be to have c++ code rely on the underlying c++ region instead of the "ScalaRegion" which is intended to just be a wrapper for the Region object in Scala.

**Comment by tpoterba** [comment]
:shipit:

**Pull request merged**